feat: no empty command completion if `no_empty_cmd_completion` is on

scop opened the pull request:
On some systems, generating the entire list of available commands takes a long time. `no_empty_cmd_completion` is semantically close enough to piggyback on it here.

Closes https://github.com/scop/bash-completion/issues/926

scop commented on test/t/unit/test_unit_compgen_commands.py:
Tests do not work but timeout for some reason, need to get that fixed up.

scop commented on bash_completion:
```diff
@@ -2557,6 +2557,15 @@ _complete_as_root()
+_comp_compgen_commands()
+{
+    [[ ! ${cur-} ]] && shopt -q no_empty_cmd_completion && return
+    _comp_compgen -- -c
```
This does not escape for example spaces in command names in `$PATH` properly. We apparently try that by setting `compopt -o filenames` in some completions (but not all), but that doesn't help.

It was similarly broken before this change too, but I'd like to either fix that while we work on this, or make it consistent if we do/don't try the filenames compopt.

scop replied:
I'm not sure how I tested before, but `compopt -o filenames` does seem to help.

akinomyoga reviewed (commented):
There are some other cases where we might suppress command name generations on empty completions.

```bash
./bash_completion:2497:        _comp_compgen -- -d -c
./completions/_su:23:            _comp_compgen -- -d -c
./completions/gdb:23:            _comp_compgen -- -d -c
```

Currently, `_comp_compgen_commands` only generates the command names, but maybe it could also generate directories by `plusdirs` by default.

scop commented:
Good points. Added `-o plusdirs` and addressed the noted remaining ones.

The note in gdb completion regarding bash builtins etc being include is valid for quite a few other things as well, that's something we may want to address sometime.

akinomyoga commented on bash_completion:
```diff
@@ -2614,6 +2612,17 @@ _complete_as_root()
+_comp_compgen_commands()
+{
+    [[ ! ${cur-} ]] && shopt -q no_empty_cmd_completion && return
+    _comp_compgen -- -c -o plusdirs
+    # for e.g. spaces in paths to and in command names
+    ((${#COMPREPLY[@]} == 0)) || compopt -o filenames
```
```suggestion
    # -o filenames for e.g. spaces in paths to and in command names
    _comp_compgen -- -c -o plusdirs && compopt -o filenames
```

For the CI error.

akinomyoga replied:
I now realized that I haven't added `@return` in the doc comment of `_comp_compgen`, but `_comp_compgen -- [options]` is intended to succeed only when at least one completion is generated, and this behavior is already used in the codebase.

It is mentioned in [api-and-naming.md in #993](https://github.com/scop/bash-completion/pull/993/files#diff-0cb3af15071817f620202dd47344251e6a8bff22af238511159fb03ef2e886f0R161-R163). I added https://github.com/scop/bash-completion/pull/993/commits/ed8e81e246d20c8ecf7f943dfe7f8e41060504ac in #993.

akinomyoga commented on bash_completion:
```diff
@@ -2614,6 +2612,17 @@ _complete_as_root()
 }
 
+# Complete on available commands, subject to `no_empty_cmd_completion`.
+#
```
I think we can define the behavior of `@return`, though there are choices.

akinomyoga replied:
```suggestion
# @return True (0) if one or more completions are generated, or otherwise False
# (1).  Note that it returns 1 even when the completion generation is canceled
# by `shopt -s no_empty_cmd_completion`.
#
```

akinomyoga commented on bash_completion:
```diff
@@ -2614,6 +2612,17 @@ _complete_as_root()
+# @since 2.12
+_comp_compgen_commands()
+{
+    [[ ! ${cur-} ]] && shopt -q no_empty_cmd_completion && return
```
If we make it the same as `_comp_compgen -- [...]` (i.e., 0 only when at least one completion is generated), we can return `1` for `no_empty_cmd_completion`.

```suggestion
    [[ ! ${cur-} ]] && shopt -q no_empty_cmd_completion && return 1
```

scop replied:
SGTM in principle, although returning 1 feels a bit "wrong" in the sense that it's not an error, it's just that the completions are inhibited because of config. There probably isn't such a use case in-tree at least yet, but a pattern like `_comp_compgen_command || do_something_else_to_generate_completions` doesn't seem completely unrealistic, but we might not want to enter the "or" branch when the nonzero is due to config. In that sense, returning 0 here could be worth considering.

But I guess it could be argued the other way in favor of a nonzero. And that'd be kind of nice for consistency, too. No strong feelings on this one here.

akinomyoga replied:
I have already let `_comp_split` and `_comp_compgen -- [...]` use exit status 1 when no completions are generated although it's not an error. Usage error is notified by status 2. A typical use case for `return 1` that I assume is `_comp_compgen -v command && process "${arr[@]}"`. But if there is another use case, there's no problem in returning 0 here.

I guess whether to run the "or" branch depends on the actual action in `do_something_elese_to_generate_completions`. For example, if `do_something_elese_to_generate_completions` has a small computational cost, I think it can be run even if `_comp_compgen_command` is canceled because the purpose of `no_empty_cmd_completion` is not to block the completion.
